File: ghost-env/src/recording/recorder.ts
export interface CallRecord {
  id: string;
  provider: string;
  method: string;
  url: string;
  requestBody?: unknown;
  responseStatus: number;
  responseBody?: unknown;
  durationMs: number;
}

export class Recorder {
  private seq = 0;
  private readonly calls: CallRecord[] = [];

  record(r: Omit<CallRecord, "id">): void {
    this.seq += 1;
    this.calls.push({ ...r, id: `c${this.seq}` });
  }

  all(): CallRecord[] {
    return [...this.calls];
  }

  filter(provider?: string): CallRecord[] {
    if (!provider) return this.all();
    return this.calls.filter((c) => c.provider === provider);
  }

  clear(): void {
    this.calls.length = 0;
    this.seq = 0;
  }
}

File: ghost-env/python/src/ghost_env/recorder.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass
class CallRecord:
    id: str
    provider: str
    method: str
    url: str
    request_body: Any
    response_status: int
    response_body: Any
    duration_ms: float


class Recorder:
    def __init__(self) -> None:
        self._calls: list[CallRecord] = []
        self._seq = 0

    def record(self, **kwargs: Any) -> None:
        self._seq += 1
        self._calls.append(CallRecord(id=f"c{self._seq}", **kwargs))

    def all(self) -> list[CallRecord]:
        return list(self._calls)

    def filter(self, provider: str | None = None) -> list[CallRecord]:
        if not provider:
            return self.all()
        return [c for c in self._calls if c.provider == provider]

    def clear(self) -> None:
        self._calls.clear()
        self._seq = 0
